### PR TITLE
dts: nrf53: Add UICR node only for configuration

### PR DESCRIPTION
**Pull request by Vge0rge**
Add a UICR node in the dts file of the non-secure
target for the NRF5340 board. This will be used solely for configuration purposes. The non-secure application cannot access the UICR and thus the register attribute is removed.

### CODE REVIEW

**Review by gmarull — CHANGES_REQUESTED** [review]
Can you please explain why we need this? There's already a UICR node.

**Comment by Vge0rge** [comment]
> Can you please explain why we need this? There's already a UICR node.

There is a uicr node in the secure only target but not in the non-secure target, right?

**Comment by gmarull** [comment]
> > Can you please explain why we need this? There's already a UICR node.
> 
> There is a uicr node in the secure only target but not in the non-secure target, right?

So what is the point of such "software" node? If you can't access UICR, how are you gonna configure NFCT pins as gpios?

**Comment by Vge0rge** [comment]
> > > Can you please explain why we need this? There's already a UICR node.
> > 
> > 
> > There is a uicr node in the secure only target but not in the non-secure target, right?
> 
> So what is the point of such "software" node? If you can't access UICR, how are you gonna configure NFCT pins as gpios?

The actual GPIO configuration will happen in the MDK inside the TF-M image. What this does is allowing us to have the same configuration mechanism for both secure and non-secure targets. It indicates to the TF-M image that this configuration should happen.

**Comment by gmarull** [comment]
> > > > Can you please explain why we need this? There's already a UICR node.
> > > 
> > > 
> > > There is a uicr node in the secure only target but not in the non-secure target, right?
> > 
> > 
> > So what is the point of such "software" node? If you can't access UICR, how are you gonna configure NFCT pins as gpios?
> 
> The actual GPIO configuration will happen in the MDK inside the TF-M image. What this does is allowing us to have the same configuration mechanism for both secure and non-secure targets. It indicates to the TF-M image that this configuration should happen.

So why is DT involved here? If TF-M can read/write to UICR, why isn't UICR then exposed to all targets? Am I missing something?

**Comment by Vge0rge** [comment]
> > > > > Can you please explain why we need this? There's already a UICR node.
> > > > 
> > > > 
> > > > There is a uicr node in the secure only target but not in the non-secure target, right?
> > > 
> > > 
> > > So what is the point of such "software" node? If you can't access UICR, how are you gonna configure NFCT pins as gpios?
> > 
> > 
> > The actual GPIO configuration will happen in the MDK inside the TF-M image. What this does is allowing us to have the same configuration mechanism for both secure and non-secure targets. It indicates to the TF-M image that this configuration should happen.
> 
> So why is DT involved here? If TF-M can read/write to UICR, why isn't UICR then exposed to all targets? Am I missing something?

The DT is involved here because since Zephyr 3.5 we suggest to the users to configure this option through DT:
https://docs.zephyrproject.org/latest/releases/migration-guide-3.5.html#recommended-changes

We use one device tree file for each build, even though when we build for the non-secure target we technically build two targets. One is the non-secure Zephyr application and the other one is TF-M (which is not Zephyr application). So the device tree file for the non-secure target is used by  the Zephyr non-secure target. This image doesn't have access to the UICR so including the UICR node with it's register address is both not useful and error-prone.

One the other hand since we now promote the use of DT for this specific option I thought that we can have a UICR node in the non-secure target for the sole purpose of configuration. This allow us to use the same configuration strategy for both targets and it is not error-prone since accessing the UICR from the Zephyr application from the  non-secure target  will result to a build failure.

**Review by nordicjm — CHANGES_REQUESTED** [review]
This looks like a hack, if it can't be read then why is it here and if it can be read why can't it use the original binding?

File: dts/arm/nordic/nrf5340_cpuappns.dtsi [hunk 1]
```diff
@@ -72,6 +72,11 @@
 		compatible = "zephyr,psa-crypto-rng";
 		status = "okay";
 	};
+
+	uicr: uicr {
+		compatible = "nordic,nrf-uicr-config-only";
+		status = "okay";
+	};
 };
```
Comment (de-nordic) [inline dts/arm/nordic/nrf5340_cpuappns.dtsi:80]: As far as I can tell from the code here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/soc_flash_nrf.c
uicr is sub-device of flash controller. That is my first issue.
My second issue: as far as I understand "compatible" describes a device/driver the node is compatible to; the usage here is hacky.

**Review by de-nordic — CHANGES_REQUESTED** [review]
No, https://github.com/zephyrproject-rtos/zephyr/pull/69337#discussion_r1504054802

**Comment by Vge0rge** [comment]
This was indeed a hack to allow users to have easier configuration. 
I fully understand that configuring this option through DTS is a good engineering solution but we need to also keep in mind the side-effects that our customers will face before deprecating Kconfig options or features in general.

The deprecation of this Kconfig makes configuration for our TF-M users more complicated since we will have to use a separate option just for TF-M and thus they have to use two separate configuration options depending on the target that they are building.

**Pull request closed without merging**